crates/providers/src/new.rs

```diff
@@ -96,7 +130,87 @@ pub trait Provider<N: Network, T: Transport + Clone = BoxTransport>: Send + Sync
     /// Returns a [`Weak`] RPC client used to send requests.
     fn weak_client(&self) -> WeakClient<T>;
 
-    async fn new_pending_transaction(&self, tx_hash: B256) -> TransportResult<PendingTransaction>;
+    /// Watch for the confirmation of a single pending transaction with the given configuration.
+    ///
+    /// Note that this is handled internally rather than calling any specific RPC method.
+    async fn watch_pending_transaction(
+        &self,
+        config: PendingTransactionConfigInner,
+    ) -> TransportResult<PendingTransaction>;
+
+    /// Notify the provider that we are interested in new blocks.
+    ///
+    /// Returns the ID to use with [`eth_getFilterChanges`](Self::get_filter_changes).
+    ///
+    /// See also [`watch_blocks`](Self::watch_blocks) to configure a poller.
+    async fn new_block_filter(&self) -> TransportResult<U256> {
+        self.client().prepare("eth_newBlockFilter", ()).await
+    }
+
+    /// Notify the provider that we are interested in new blocks.
+    ///
+    /// Returns the ID to use with [`eth_getFilterChanges`](Self::get_filter_changes).
+    ///
+    /// See also [`watch_blocks`](Self::watch_blocks) to configure a poller.
+    async fn new_pending_transactions_filter(&self) -> TransportResult<U256> {
+        self.client().prepare("eth_newPendingTransactionFilter", ()).await
+    }
+
+    /// Notify the provider that we are interested in new logs using the given filter.
+    ///
+    /// Returns the ID to use with [`eth_getFilterChanges`](Self::get_filter_changes).
+    ///
+    /// See also [`watch_logs`](Self::watch_logs) to configure a poller.
+    async fn new_filter(&self, filter: &Filter) -> TransportResult<U256> {
+        self.client().prepare("eth_newFilter", (filter,)).await
+    }
+
+    /// Get a list of values that have been added since the last poll.
+    ///
+    /// The return value depends on what stream `id` corresponds to.
+    /// See [`FilterChanges`] for all possible return values.
+    #[auto_impl(keep_default_for(&, &mut, Rc, Arc, Box))]
```

Comment:
TIL this is necessary for `dyn Provider` to work, because otherwise `auto_impl` puts `U: Provider + Sized` bounds on the generic implementations if `Self: Sized` methods are not marked with `keep_default_for`. I added a test for this in this file